source/dotnet/Application/Batches/BatchApplicationService.cs

```
@@ -22,17 +23,52 @@ public class BatchApplicationService : IBatchApplicationService
 {
     private readonly IBatchRepository _batchRepository;
     private readonly IUnitOfWork _unitOfWork;
+    private readonly IBatchRunner _batchRunner;
+    private readonly IProcessCompletedPublisher _processCompletedPublisher;
 
-    public BatchApplicationService(IBatchRepository batchRepository, IUnitOfWork unitOfWork)
+    public BatchApplicationService(IBatchRepository batchRepository, IUnitOfWork unitOfWork, IBatchRunner batchRunner, IProcessCompletedPublisher processCompletedPublisher)
     {
         _batchRepository = batchRepository;
         _unitOfWork = unitOfWork;
+        _batchRunner = batchRunner;
+        _processCompletedPublisher = processCompletedPublisher;
     }
 
     public async Task CreateAsync(BatchRequestDto batchRequestDto)
     {
-        var batch = new Batch(batchRequestDto.ProcessType, batchRequestDto.GridAreaCodes.Select(c => new GridAreaCode(c)));
+        var gridAreaCodes = batchRequestDto.GridAreaCodes.Select(c => new GridAreaCode(c));
+        var batch = new Batch(batchRequestDto.ProcessType, gridAreaCodes);
+
         await _batchRepository.AddAsync(batch).ConfigureAwait(false);
         await _unitOfWork.CommitAsync().ConfigureAwait(false);
     }
+
+    public async Task StartRequestedAsync()
+    {
+        var requestedBatches = await _batchRepository.GetRequestedAsync().ConfigureAwait(false);
+        await _batchRunner.BeginExecuteAsync(requestedBatches).ConfigureAwait(false);
```

Case: requestedBatches contains 7 pending batches. BatchRunner is able to start the first 4 batches and then fails with an exception. How can the application recover from this invalid state where 4 jobs have started but the state is not reflected?

---

I anticipate that when getting there we might do something like this:
Let `BatchRunner.BeginExecuteAsync()` first check if the batches have actually already been started (and possibly finished). And in that case, silently ignore them. Then they will get the "executing" state and eventually be found completed.

Do you have different thoughts?

---

I would try to break it down and process each batch individually. Handle each batch in it's own unit-of-work.

---

The offending code has been removed.